packages/devtools_flutter/lib/src/config.dart

```diff
@@ -10,28 +11,111 @@ import 'screen.dart';
+class DevToolsAppState extends State<DevToolsApp> {
+  @override
+  void initState() {
+    super.initState();
+    service.FrameworkCore.init(WidgetsBinding.instance.window.defaultRouteName);
+  }
+
+  /// Generates routes, separating the path from URL query parameters.
+  Route generateRoute(RouteSettings settings) {
+    final uri = Uri.parse(settings.name);
+    final queryParams = uri.queryParameters;
+    final path = uri.path;
+    print('path: $path, params: $queryParams');
+
+    if (routes.containsKey(path)) {
+      return MaterialPageRoute(settings: settings, builder: routes[path]);
     }
-    return routeToBuilder;
+    // Return a page not found.
+    return MaterialPageRoute(
+      settings: settings,
+      builder: (_) {
+        return DevToolsScaffold.withChild(
+          child: const Center(
+            child: Text('Sorry, this page was not found'),
+          ),
+        );
+      },
+    );
   }
 
-  // The mapping from routes to the collection of screens to show in the app.
-  //
-  // The /connect route will be a dependency for all the other routes.
-  final Map<String, List<Screen>> _routeToTabs = const {
-    '/': [ConnectScreen()],
-    '/connected': [
-      EmptyScreen.inspector,
-      EmptyScreen.timeline,
-      EmptyScreen.performance,
-      EmptyScreen.memory,
-      EmptyScreen.logging,
-    ],
+  /// The routes that the app exposes.
+  final Map<String, WidgetBuilder> routes = {
+    '/': (_) => Initializer(
+          builder: (_) => const DevToolsScaffold(
+            tabs: [
+              EmptyScreen.inspector,
+              EmptyScreen.timeline,
+              EmptyScreen.performance,
+              EmptyScreen.memory,
+              EmptyScreen.logging,
+            ],
+          ),
+        ),
+    '/connect': (_) => DevToolsScaffold.withChild(child: ConnectScreenBody()),
   };
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      theme: ThemeData.light(),
+      onGenerateRoute: generateRoute,
+    );
+  }
+}
+
+/// Widget that requires business logic to be loaded before building its
+/// [builder].
+///
+/// See [_InitializerState.build] for the logic that determines whether the
+/// business logic is loaded.
+///
+/// Use this widget to wrap pages that require [service.serviceManager] to be
+/// connected. As we require additional services to be available, add them
+/// here.
+class Initializer extends StatefulWidget {
+  const Initializer({Key key, @required this.builder})
+      : assert(builder != null),
+        super(key: key);
+
+  /// The builder for the widget's children.
+  ///
+  /// Will only be built if [_InitializerState._loaded] is true.
+  final WidgetBuilder builder;
+
+  @override
+  _InitializerState createState() => _InitializerState();
+}
+
+class _InitializerState extends State<Initializer> {
+  @override
+  Widget build(BuildContext context) {
+    final loaded = service.serviceManager.hasConnection;
```

Comment:
I'm not seeing how this method waits for business logic to occur before building the widget. Nothing is being awaited, so it seems like we do a one time check for whether or not the vm service has a connection, and then load the connect screen if not.

what if we connect to the vm service immediately after `loaded` is set? Seems like this could result in a race condition where we think we aren't connected to the vm service and we actually are. Why not just `await service.serviceManager.serviceAvailable` instead of checking for `service.serviceManager.hasConnection`?

Reply:
Widget build methods can't be async.  We have to return a widget, and we have to return it reasonably quickly to avoid producing lag.

That said, we could use the serviceAvailable API from outside the build method.  I'll look into this.